python_scripts/01_adult_census_exploration.py

```diff
@@ -200,6 +199,53 @@
 # this may be a factor when deciding which model to chose.
 
 
+# %%
+def plot_tree_decision_function(tree, X, y, ax):
+    import numpy as np
+    from scipy import ndimage
+
+    h = 0.02
+    x_min, x_max = X.iloc[:, 0].min() - .5, X.iloc[:, 0].max() + .5
+    y_min, y_max = X.iloc[:, 1].min() - .5, X.iloc[:, 1].max() + .5
+    xx, yy = np.meshgrid(np.arange(x_min, x_max, h),
+                         np.arange(y_min, y_max, h))
+
+    Z = tree.predict_proba(np.c_[xx.ravel(), yy.ravel()])[:, 1]
+    Z = Z.reshape(xx.shape)
+    faces = tree.tree_.apply(
+        np.c_[xx.ravel(), yy.ravel()].astype(np.float32)
+    )
+    faces = faces.reshape(xx.shape)
+    border = ndimage.laplace(faces) != 0
+    ax.scatter(
+        X.iloc[:, 0], X.iloc[:, 1], c=np.array(['tab:blue', 'tab:red'])[y],
+        s=60, alpha=0.7
+    )
+    ax.contourf(xx, yy, Z, alpha=.4, cmap='RdBu_r')
+    ax.scatter(xx[border], yy[border], marker='.', s=1)
+    ax.set_xlabel(X.columns[0])
+    ax.set_ylabel(X.columns[1])
+
+
+# %%
+import matplotlib.pyplot as plt
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.tree import plot_tree
+from sklearn.preprocessing import LabelEncoder
+
+data_subset = adult_census[:n_samples_to_plot]
+X = data_subset[["age", "hours-per-week"]]
+y = LabelEncoder().fit_transform(
+    data_subset[target_column].to_numpy()
+)
+
+max_depth = 2
+tree = DecisionTreeClassifier(max_depth=max_depth, random_state=0)
```

Comment:
Please use `max_leaf_nodes=3` instead of `max_depth`. This way we get the 2 purish leafs found manually + the non-pure, difficult to classify area.